packages/workbox-cache-expiration/CacheExpirationManager.mjs

```diff
@@ -0,0 +1,160 @@
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+import {_private} from 'workbox-core';
+import core from 'workbox-core';
+
+import CacheTimestampsModel from './models/CacheTimestampsModel.mjs';
+
+/**
+ * The `CacheExpiration` class allows you define an expiration and / or
+ * limit on the number of responses stored in a
+ * [`Cache`](https://developer.mozilla.org/en-US/docs/Web/API/Cache).
+ */
+class CacheExpirationManager {
+  /**
+   * To construct a new CacheExpiration instance you must provide at least
+   * one of the `config` properties.
+   *
+   * @param {string} cacheName Name of the cache to apply restrictions to.
+   * @param {Object} config
+   * @param {number} [config.maxEntries] The maximum number of entries to cache.
+   * Entries used the least will be removed as the maximum is reached.
+   * @param {number} [config.maxAgeSeconds] The maximum age of an entry before
+   * it's treated as stale and removed.
+   */
+  constructor(cacheName, config) {
+    if (process.env.NODE_ENV !== 'production') {
+      core.assert.isType(cacheName, 'string', {
+        moduleName: 'workbox-cache-expiration',
+        className: 'CacheExpirationManager',
+        funcName: 'constructor',
+        paramName: 'cacheName',
+      });
+
+      if (!(config.maxEntries || config.maxAgeSeconds)) {
+        throw new _private.WorkboxError('max-entries-or-age-required');
+      }
+
+      if (config.maxEntries) {
+        core.assert.isType(config.maxEntries, 'number', {
+          moduleName: 'workbox-cache-expiration',
+          className: 'CacheExpirationManager',
+          funcName: 'constructor',
+          paramName: 'config.maxEntries',
+        });
+      }
+
+      if (config.maxAgeSeconds) {
+        core.assert.isType(config.maxAgeSeconds, 'number', {
+          moduleName: 'workbox-cache-expiration',
+          className: 'CacheExpirationManager',
+          funcName: 'constructor',
+          paramName: 'config.maxAgeSeconds',
+        });
+      }
+    }
+
+    this._maxEntries = config.maxEntries;
+    this._maxAgeSeconds = config.maxAgeSeconds;
+    this._timestampModel = new CacheTimestampsModel(cacheName);
+  }
+
+  /**
+   * Expires entries for the given cache and given criteria.
+   */
+  async expireEntries() {
+    const now = Date.now();
+
+    // First, expire old entries, if maxAgeSeconds is set.
+    const oldEntries = await this._findOldEntries(now);
+
+    // Once that's done, check for the maximum size.
+    const extraEntries = await this._findExtraEntries();
+
+    // Use a Set to remove any duplicates following the concatenation, then
+    // convert back into an array.
+    const urls = [...new Set(oldEntries.concat(extraEntries))];
+    await Promise.all([
+      this._deleteFromCache(urls),
+      this._deleteFromIDB(urls),
+    ]);
+
+    if (process.env.NODE_ENV) {
```

Comment:
Is this meant to be `process.env.NODE_ENV !== 'production'`? If not, when would this be false?

Reply:
Good spot - fixed hopefully.